### ISSUE TITLE
Enhance Results page with sorting

### ISSUE THREAD

**Issue by PatReynolds** [FreeUKGen/FreeBMD2#268]
The previous version of the Restults page had separate columns for first name, last name, date, etc. which could be used to sort the results. This is something new to FreeBMD2 (not possible on FreeBMD1).  Now there are only three columns, sorting is no longer possible.

Options might be:
 * buttons to sort by first name, sort by last name, etc.
 * button that says 'sort' and offers the choices of first name, last name, etc.
 * small buttons in the header that have a distinguishing shape, letter, colour etc and an info button that says "sort by first name", 'sort by last name", etc
 * put the columns back

Ad sorting is not possible on FreeBMD1, this is not strictly MVP - by is a nice 'extra' to give to those wanting to know what is _better_ about FreeBMD2.

**Comment by PatReynolds** [comment]
@KatyMawhood and @Daniel-silve1ra to take a look - also to consider whether implementation should be MVP or v.5

**Comment by PatReynolds** [comment]
Suggest: have a  dropdown for sort by date/district/first name/lastname/ (in that order).
Place sort dropdown before filter and show dropdowns ('show' dropdown can be much narrower than it is currently, to gain some space.

**Comment by PatReynolds** [comment]
Functionality to be discussed by Pat/Denise/Alison at next documentation meeting

**Comment by DeniseColbert** [comment]
As per comment 25/9 plus remove sort function from column header:

![image](https://user-images.githubusercontent.com/19173257/141469481-298079e0-cf23-4fa2-a886-27e1d6524e8c.png)

**Comment by PatReynolds** [comment]
Discussed in Documentation Meeting 12 Nov 2021

**Comment by PatReynolds** [comment]
@Vino-S to provide 'sort by' button and remove sort from column headings, as described above

**Comment by Vino-S** [comment]
Ready for testing in test2

**Comment by AlOneill** [comment]
Cannot open Tools drop-down with either space or enter (the expected keys).

**No visible warning** for someone who cannot hover that District link will open in a new tab.

Sorting fields are keyboard-accessible, using space-bar — functions not fully tested, but I note that sort by First Name(s) works as expected.

**Comment by DeniseColbert** [comment]
Order is currently:

![image](https://user-images.githubusercontent.com/19173257/144891616-76dc2165-b6ba-460f-a195-480d9d7e0c2a.png)

But should be:

date/district/first name/lastname

**Comment by DeniseColbert** [comment]
Record Type isn't required in this menu (as it's addressed in the Filter By menu).

**Comment by Vino-S** [comment]
Ready for testing in test3

**Comment by richpomfret** [comment]
Review/test the logic with the sorting and default ascending/descending.

**Comment by PatReynolds** [comment]
Sort by date (as per BMD1), revisit this story for research and possible enhancement in a later phase.

**Comment by DeniseColbert** [comment]
Sort order is currently latest to earliest but should be earliest to latest per FreeBMD1

**Comment by Vino-S** [comment]
Ready for testing in test3

**Comment by PatReynolds** [comment]
@PatReynolds to test

**Comment by PatReynolds** [comment]
sort by district failed
![image.png](https://images.zenhubusercontent.com/5b695a5dac3a200155c765e3/16493d9c-69f8-4a59-b9f5-755d5b532358)

sort by first name sorted Z-A, A-Z would be expected. Pressed 'update results" again and sorted A-Z

Sort by last name also sorted Z-A first.  I like that it also subsorts by first name. Search used was:
![image.png](https://images.zenhubusercontent.com/5b695a5dac3a200155c765e3/2cc8d366-10e7-4c8d-a780-29420e846aee)

**Comment by richpomfret** [comment]
@Vino-S please take a look and fix when possible.

**Comment by richardofsussex** [comment]
I think I've fixed the sorting issues, both Z-A and districts. While I was in there I added volume and page to both the search results listing and the CSV output.
Code changes are with @Vino-S for review and deployment to test3.

**Comment by richpomfret** [comment]
Ready for testing on test3

**Comment by DeniseColbert** [comment]
Search performed: Jane Partington (nothing else) with 523 results.

Sort by district returned this:
![image](https://user-images.githubusercontent.com/19173257/188611483-cc242e2e-ab53-4a7c-afbd-331b7efce47c.png)

But sort by first name returned more Janeys than are in the first snip
![image](https://user-images.githubusercontent.com/19173257/188612256-32b47510-f55f-4b69-b03b-1290e58ff3be.png)


I think this may be another instance of the Sentence Case/UPPER CASE issue...

**Comment by richardofsussex** [comment]
The "Sort by District" option seems to be sorting, in fact, by first name. It's certainly not sorting by District. There are subtleties about whether it is sorting by surname as well, but this is secondary. Ideally the sort should force everything to upper or lower case before comparing it, so that Partington and PARTINGTON are treated as equivalent.

Sorting clearly needs more work.

**Comment by richardofsussex** [comment]
OK, I have fixed the failure to sort by District. It was trying to use the non-existent DistrictName field! Also made the sorting case-insensitive, so that PARTINGTON and Partington should sort as "the same".

**Comment by DeniseColbert** [comment]
Looks much better/fixed. However, how are results with the same name and district sorted? You'd expect by date, but that doesn't seem to be the case, e.g. same search, sort by district, page 2:

![image](https://user-images.githubusercontent.com/19173257/189655135-81fb2fd4-0001-4a43-b9b9-485b5a2361e3.png)

The three Barton results here, one would expect them to be sorted from earliest to latest.

**Comment by richardofsussex** [comment]
I think this is a consequence of implementing case-insensitivity. I do this by forcing each of the fields to be compared to a string, then lower-casing these strings. However, the QuarterNumber field is an integer. If you look at the results sorted by date, they are clearly wrong (check the last page). I will have another go!

**Comment by richardofsussex** [comment]
OK, fb_268_updated_rbl contains updated code which only lower-cases string fields when comparing them. Please deploy to test3 and check.

**Comment by Vino-S** [comment]
Hi @richardofsussex , test3 is updated

**Comment by richardofsussex** [comment]
Still doesn't work! Sadly, my test file is too small to test sorting of QuarterNumber, so I have to push to test3 to see if my changes work. I'll have another go.

**Comment by richardofsussex** [comment]
OK, @Vino-S, fb_268_updated_rbl contains updated code which only lower-cases string fields when comparing them. Please deploy to test3 and check.

**Comment by richardofsussex** [comment]
Fixed a bug in case-insensitive multi-field sorting. Should be OK now.

**Comment by DeniseColbert** [comment]
To be checked after deployment

**Comment by DeniseColbert** [comment]
@DeniseColbert  to check on beta (& test3)

**Comment by DeniseColbert** [comment]
Vino has tested on test 3 and working, Denise to test beta.

**Comment by DeniseColbert** [comment]
Appears fixed, thanks all!

**Comment by DeniseColbert** [comment]
Closing
